src/learning/gmm.rs

```diff
@@ -326,11 +338,30 @@ impl GaussianMixtureModel {
             new_covs.push(cov_mat / sum_weights[k]);
 
         }
-
+        self.bic = self.calculate_bic(samples);
         self.model_means = Some(new_means);
         self.model_covars = Some(new_covs);
     }
 
+    ///Calculates the model's Bayesian Information Criterion (BIC)
+    /// BIC = -2*log(l) + k * ln(n)
+    /// useful for determining the optimal number of clusters when iteratively generating GMMs.
+    /// log_lik = log likelihood criterion for the model, the calcaulated log_lik parameter is a sum so it needs to be divided by the total number of samples.
+    /// num_clusters = the number of clusters created in the model.
+    /// n = the total number of samples used to create the model.
+    fn calculate_bic(&self, n: f64) -> f64 {
+        let num_clusters:f64 = self.comp_count as f64;
+        let log_lik:f64 = self.log_lik / n;
+        let log_samples:f64 = n.ln();
+        assert!(!num_clusters.is_nan());
```

Comment:
I think it is better that these are debug assertions (`debug_assert!`) - especially as currently they go in the inner loop.